contracts/governance/extensions/GovernorProposalGuardian.sol

```diff
@@ -0,0 +1,75 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import {Governor} from "../Governor.sol";
+
+/**
+ * @dev Extension of {Governor} which adds a proposal guardian that can cancel proposals at any stage of their lifecycle.
+ *
+ * Note: if the proposal guardian is not configured, then proposers take this role for their proposals.
+ */
+abstract contract GovernorProposalGuardian is Governor {
+    address private _proposalGuardian;
+
+    event ProposalGuardianSet(address oldProposalGuardian, address newProposalGuardian);
+
+    /**
+     * @dev Override {IGovernor-cancel} that implements the extended cancellation logic.
+     * * proposal guardian can cancel any proposal at any point in the lifecycle.
+     * * if no proposal guardian is set, proposer can cancel their proposals at any point in the lifecycle.
+     * * if the proposal guardian is set, proposer keep their ability to cancel during the pending stage (default behavior).
+     */
+    function cancel(
+        address[] memory targets,
+        uint256[] memory values,
+        bytes[] memory calldatas,
+        bytes32 descriptionHash
+    ) public virtual override returns (uint256) {
+        address caller = _msgSender();
+        address authority = proposalGuardian();
+
+        if (authority == address(0)) {
+            // if there is no proposal guardian
+            // ... only the proposer can cancel
+            // ... no restriction on when the proposer can cancel
+            uint256 proposalId = getProposalId(targets, values, calldatas, descriptionHash);
+            address proposer = proposalProposer(proposalId);
+            if (caller != proposer) revert GovernorOnlyProposer(caller);
+            return _cancel(targets, values, calldatas, descriptionHash);
```

Comment:
The default behavior in Governor is to only allow the proposer to cancel when the proposal is still pending. I don't think we want to allow proposers to cancel after that.

Also, if we want to keep the default behavior, wouldn't it be better to just?:

```suggestion
            return super.cancel(targets, values, calldatas, descriptionHash);
```

This would simplify this function imo